pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+// startupMonitorPodConditionController monitors the startup monitor pod and reports unwanted state/condition/phase as a degraded condition
+type startupMonitorPodConditionController struct {
+	operatorClient operatorv1helpers.OperatorClient
+
+	targetName string
+	podLister  corev1listers.PodNamespaceLister
+
+	startupMonitorEnabledFn func() bool
+}
+
+// New returns a controller for monitoring the lifecycle of the startup monitor pod
+func New(targetNamespace string,
+	targetName string,
+	operatorClient operatorv1helpers.OperatorClient,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	startupMonitorEnabledFn func() bool,
+	eventRecorder events.Recorder) factory.Controller {
+	fd := &startupMonitorPodConditionController{
+		operatorClient:          operatorClient,
+		targetName:              targetName,
+		podLister:               kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Lister().Pods(targetNamespace),
+		startupMonitorEnabledFn: startupMonitorEnabledFn,
+	}
+	return factory.New().WithSync(fd.sync).ResyncEvery(6*time.Minute).WithInformers(kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Informer()).ToController("StartupMonitorPodCondition", eventRecorder)
+}
+
+func (fd *startupMonitorPodConditionController) sync(_ context.Context, _ factory.SyncContext) (err error) {
+	startupPodDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodDegraded", Status: operatorv1.ConditionFalse}
+	startupPodContainerExcessiveRestartsDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodContainerExcessiveRestartsDegraded", Status: operatorv1.ConditionFalse}
+
+	defer func() {
+		if err == nil {
+			if _, _, updateError := operatorv1helpers.UpdateStatus(fd.operatorClient,
+				operatorv1helpers.UpdateConditionFn(*startupPodDegraded),
+				operatorv1helpers.UpdateConditionFn(*startupPodContainerExcessiveRestartsDegraded)); updateError != nil {
+				err = updateError
+			}
+		}
+	}()
+
+	// in practice we rely on operators to provide
+	// a condition for checking we are running on a single node cluster
+	if !fd.startupMonitorEnabledFn() {
+		return nil
+	}
+
+	monitorPod, err := fd.podLister.Get(fmt.Sprintf("%s-startup-monitor", fd.targetName))
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err // unknown error
+	}
+
+	withDegradedWhenInPendingPhaseFor(monitorPod, 2*time.Minute, startupPodDegraded)
+	withDegradedWhenHasFailedContainer(monitorPod, startupPodDegraded)
+	withDegradedWhenHasExcessiveRestarts(monitorPod, 2, startupPodContainerExcessiveRestartsDegraded)
```

Comment:
is it possible that `withDegradedWhenHasExcessiveRestarts` shadows some of the previous?

Reply:
I mean, should any of `withDegradedWhen*`return immediately once if finds a problem?

Reply:
I don't think so, hence a new condition. The previous condition will be set when a pod failed with a non-zero exit code or has been pending for more than 2 minutes.